File: src/pratice/graph/bfs/단지번호붙이기2667.java
package pratice.graph.bfs;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class 단지번호붙이기2667 {
    public static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    public static int N;
    public static int[][]A;
    public static boolean[][]visited;
    public static int[] dx = new int[]{1, 0, -1, 0};
    public static int[] dy = new int[]{0, 1, 0, -1};

    public static void main(String[] args) throws IOException {
        input();
        pro();
    }

    private static void pro() {
        int cnt = 0;
        ArrayList<Integer> result = new ArrayList<>();
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                if(A[i][j] != 0 && !visited[i][j]){
                    cnt++;
                    int total = bfs(new Point(i, j));
                    result.add(total);
                }
            }
        }
        System.out.println(cnt);
        result.sort(Comparator.comparingInt(a -> a));
        for (int i = 0; i < result.size(); i++) {
            System.out.println(result.get(i));
        }
    }

    private static int bfs(Point point) {
        Queue<Point> needVisited = new LinkedList<>();
        needVisited.add(point);
        int total = 0;
        while(!needVisited.isEmpty()){
            Point p = needVisited.poll();
            if(!visited[p.x][p.y]){
                visited[p.x][p.y] = true;
                total++;
                for (int i = 0; i < 4; i++) {
                    int x = p.x + dx[i];
                    int y = p.y + dy[i];
                    if(x >= 0 && y >= 0 && x < N && y < N){
                        if(!visited[x][y] && A[x][y] != 0){
                            needVisited.add(new Point(x, y));
                        }
                    }
                }
            }
        }
        return total;
    }
    private static void input() throws IOException {
        N = Integer.parseInt(br.readLine());
        A = new int[N][N];
        visited = new boolean[N][N];
        for (int i = 0; i < N; i++) {
            char[] chars = br.readLine().toCharArray();
            for (int j = 0; j <N ; j++) {
                A[i][j] = chars[j] - '0';
            }
        }
    }
}
class Point{
    public int x;
    public int y;

    public Point(int x, int y) {
        this.x = x;
        this.y = y;
    }
}

File: src/pratice/graph/bfs/단지번호붙이기2667.kt
package pratice.graph.bfs

import java.util.*

private var N: Int = 0
private val dx = arrayOf(0, 1, 0, -1)
private val dy = arrayOf(1, 0, -1, 0)
private lateinit var arr: Array<IntArray>
private lateinit var visited: Array<BooleanArray>
private val blockSizeList = mutableListOf<Int>()
private var totalBlockSize = 0
fun main() {
    N = readln().toInt()
    arr = Array(N) { IntArray(N) }
    visited = Array(N) { BooleanArray(N) }

    for (i in 0 until N) {
        arr[i] = readln().toCharArray().map { it - '0' }.toIntArray()
    }
    for (i in 0 until N) {
        for (j in 0 until N) {
            if (arr[i][j] == 1 && !visited[i][j]) {
                totalBlockSize++
                bfs(i, j)
            }
        }
    }
    blockSizeList.sort()
    println(totalBlockSize)
    blockSizeList.forEach { println(it) }
}

private fun bfs(startX: Int, startY: Int) {
    val needVisited = LinkedList<Int>()
    needVisited.add(startX)
    needVisited.add(startY)
    var blockSize = 0
    while (needVisited.isNotEmpty()) {
        val px = needVisited.poll()
        val py = needVisited.poll()
        if (visited[px][py]) {
            continue
        }
        visited[px][py] = true
        blockSize++
        for (i in 0 until 4) {
            val x = px + dx[i]
            val y = py + dy[i]
            if (x >= 0 && y >= 0 && x < N && y < N) {
                if (!visited[x][y] && arr[x][y] == 1) {
                    needVisited.add(x)
                    needVisited.add(y)
                }
            }
        }
    }
    blockSizeList.add(blockSize)
}
